Intrinsic memset optimization

rafaelsamenezes opened the pull request:
This PR will optimize memset operations by initializing the entire object with the result of the memset. This should avoid unrolling and speed up for memset of known bounds.

Closes #635 

We should review this carefully, as this is a critical part of esbmc. I will also run this on sv-benchmarks.

rafaelsamenezes committed:
[builtin] memset now generates an object

In special, added a method to generate a memset by bitwise operations

rafaelsamenezes committed:
[memset] added testcases for memset builtin

lucasccordeiro commented on src/goto-symex/builtin_functions.cpp:
```diff
@@ -748,277 +748,389 @@ void goto_symext::symex_va_arg(const expr2tc &lhs, const sideeffect2t &code)
-void goto_symext::intrinsic_memset(
-  reachability_treet &art,
-  const code_function_call2t &func_call)
+// Computes the equivalent object value when considering a memset operation on it
```
Can you also document this method in the .h file?

lucasccordeiro commented on src/goto-symex/builtin_functions.cpp:
```diff
@@ -748,277 +748,389 @@ void goto_symext::symex_va_arg(const expr2tc &lhs, const sideeffect2t &code)
+   * in the case where every byte `value` was set set up until num_of_bytes
+   * 
+   * @warning this function does not add any pointer/memory/bounds check!
+   *          they should be added before calling this function!
```
How could we do it?

lucasccordeiro commented on src/goto-symex/builtin_functions.cpp:
```diff
@@ -748,277 +748,389 @@ void goto_symext::symex_va_arg(const expr2tc &lhs, const sideeffect2t &code)
+   *    and then, until the num_of_bytes is reached it will do a full byte
+   *    left-shift followed by an bitor operation with the byte value:
+   * 
+   *    Example, for a integer(4 bytes) with memset using 3 bytes and value 0xF1
```
"Example" -> "For example"

lucasccordeiro commented on src/goto-symex/builtin_functions.cpp:
```diff
@@ -748,277 +748,389 @@ void goto_symext::symex_va_arg(const expr2tc &lhs, const sideeffect2t &code)
+   * 
+   * B. Generate a mask of the bits that were not set, this is done because skipped bits
+   *    need to be returned back. The computation of this is simple, we initialize every
+   *    bit that was changed by the byte-representation computation with a 1. Which is then
```
"with a 1. Which is then" -> "with a 1, which is then"

lucasccordeiro commented on src/goto-symex/builtin_functions.cpp:
```diff
@@ -748,277 +748,389 @@ void goto_symext::symex_va_arg(const expr2tc &lhs, const sideeffect2t &code)
+  auto value_downcast = typecast2tc(get_uint8_type(), value);
+  auto value_upcast = typecast2tc(
+    type,
+    value_downcast); // so smt_conv will complain about the width of the type
```
Why? Could we refactor the code?

mikhailramalho commented on src/goto-symex/builtin_functions.cpp:
```diff
@@ -748,277 +748,389 @@ void goto_symext::symex_va_arg(const expr2tc &lhs, const sideeffect2t &code)
-  reachability_treet &art,
-  const code_function_call2t &func_call)
+// Computes the equivalent object value when considering a memset operation on it
+expr2tc gen_byte_expression(
```
static inline?

mikhailramalho commented on src/goto-symex/builtin_functions.cpp:
```diff
@@ -748,277 +748,389 @@ void goto_symext::symex_va_arg(const expr2tc &lhs, const sideeffect2t &code)
-  for(const auto &item : internal_deref_items)
-  {
-    const expr2tc &offs = item.offset;
+  auto simplified = result->simplify();
```
I think we have a method simplify that can be used here instead, as it honors the --no-simplify flag.

mikhailramalho commented on src/goto-symex/builtin_functions.cpp:
```diff
@@ -748,277 +748,389 @@ void goto_symext::symex_va_arg(const expr2tc &lhs, const sideeffect2t &code)
-    {
-      continue;
-    }
+inline expr2tc gen_value_by_byte(
```
static inline?

mikhailramalho commented on src/goto-symex/builtin_functions.cpp:
```diff
@@ -748,277 +748,389 @@ void goto_symext::symex_va_arg(const expr2tc &lhs, const sideeffect2t &code)
+
+    // Is this pointing to a valid place?
+    if(!internal_deref_items.size())
+      throw std::invalid_argument("Couldn't deref the object");
```
Can we remove the throws and simply guard against these error cases? Calling the code in the catch block when the if fails.

rafaelsamenezes replied:
replaced throws with an empty expression

mikhailramalho commented on src/goto-symex/builtin_functions.cpp:
```diff
@@ -748,277 +748,389 @@ void goto_symext::symex_va_arg(const expr2tc &lhs, const sideeffect2t &code)
+
+  // I am not sure if bitwise operations are valid for floats
+  if(is_floatbv_type(type) || is_fixedbv_type(type))
+    throw std::invalid_argument("This will not work for floats");
```
can we get rid of the throws here? You can return expr2tc() and check the returned value on the caller.

rafaelsamenezes replied:
replaced throws with an empty expression

mikhailramalho reviewed (commented):
Overall it looks go, except for a few nits.

As always, I'd like to see the sv-comp results before merging, as this piece of code is quite performance critical.

rafaelsamenezes commented:
@mikhailramalho and @lucasccordeiro Thanks for the reviews! I will actually put it back as a draft, there are 2 features still missing: symbol support and multiple dereference objects. I will address your comments, implement those features and mark it as ready again.

rafaelsamenezes committed:
[memset] added function to handle symbols by byte_update

rafaelsamenezes committed:
[memset] optimizations for pointers typecasts

rafaelsamenezes committed:
[code-style] applied clang-format

rafaelsamenezes committed:
[memset] added checks

rafaelsamenezes committed:
[regression] 05_pfscan is now working correctly

rafaelsamenezes committed:
[memset] simplifications for base offset

rafaelsamenezes committed:
[memset] added check for code

rafaelsamenezes committed:
[memset] added check for constant_int

rafaelsamenezes committed:
[memset] added check for structures members

rafaelsamenezes committed:
Update test.desc

lucasccordeiro commented:
@rafaelsamenezes: What is the status of this PR?

fbrausse commented on src/goto-symex/builtin_functions.cpp:
```diff
@@ -748,18 +748,378 @@ void goto_symext::symex_va_arg(const expr2tc &lhs, const sideeffect2t &code)
+    type,
+    value_downcast); // so smt_conv will complain about the width of the type
+
+  expr2tc mask = gen_zero(type);
```
I don't understand why `mask` is generated symbolically here. The loop iterations and therefore #shifts are known statically below. Instead something like this could be done keeping the expressions smaller and the simplifier less busy:
```c++
BigInt n, o;
n.setPower2(num_of_bytes * 8); /* or config.ansi_c.char_width if we want to be pedantic */
o.setPower2(offset * 8);
mask = constant_int2tc(type, (n-1) * o);
```

rafaelsamenezes replied:
I've tried this approach and it doesn't seems to be working. Anyway, the formula is simplified. It shouldn't add much cost.

fbrausse commented on src/goto-symex/builtin_functions.cpp:
```diff
@@ -748,18 +748,378 @@ void goto_symext::symex_va_arg(const expr2tc &lhs, const sideeffect2t &code)
+  expr2tc mask = gen_zero(type);
+
+  auto eight = constant_int2tc(int_type2(), BigInt((int)8));
+  auto one = constant_int2tc(int_type2(), BigInt((int)1));
```
I believe neither the `(int)` casts nor the explicit calls to `BigInt()` are necessary here.

fbrausse commented on src/goto-symex/builtin_functions.cpp:
```diff
@@ -748,18 +748,378 @@ void goto_symext::symex_va_arg(const expr2tc &lhs, const sideeffect2t &code)
+   * 
+   */
+
+  // I am not sure if bitwise operations are valid for floats
```
They are invalid in C and C++ and I hope none of the other frontends generate those. This constraint is also part of #204.

fbrausse commented on src/goto-symex/builtin_functions.cpp:
```diff
@@ -748,18 +748,378 @@ void goto_symext::symex_va_arg(const expr2tc &lhs, const sideeffect2t &code)
+
+    for(unsigned i = 0; i < result->datatype_members.size(); i++)
+    {
+      auto name = to_struct_type(type).member_names[i];
```
Maybe it's just me, but the fewer uses of unnecessary `auto` the better when refactoring because the compiler will complain about all places that might need to be touched. Would you consider decreasing the usage of `auto`?

rafaelsamenezes replied:
Sure, no problem

fbrausse commented on src/goto-symex/builtin_functions.cpp:
```diff
@@ -748,18 +748,378 @@ void goto_symext::symex_va_arg(const expr2tc &lhs, const sideeffect2t &code)
+      {
+        assert(offset_left < current_member_size);
+        auto bytes_to_write =
+          bytes_left < current_member_size ? bytes_left : current_member_size;
```
`std::min`?

rafaelsamenezes replied:
ACK

fbrausse commented on src/goto-symex/builtin_functions.cpp:
```diff
@@ -780,246 +1140,181 @@ void goto_symext::intrinsic_memset(
-  dereference(deref, dereferencet::INTERNAL);
+  auto simplified = arg2->simplify();
+  if(simplified)
+    arg2 = simplified;
```
`simplify(arg2)`?

rafaelsamenezes replied:
Ow, didn't know that we could use like this.

fbrausse commented on src/goto-symex/builtin_functions.cpp:
```diff
@@ -780,246 +1140,181 @@ void goto_symext::intrinsic_memset(
+
+    auto offset_simplified = item_offset->simplify();
+    if(offset_simplified)
+      item_offset = offset_simplified;
```
same: `simplify(...)`?

rafaelsamenezes replied:
Same as above

fbrausse commented on src/goto-symex/builtin_functions.cpp:
```diff
@@ -780,246 +1140,181 @@ void goto_symext::intrinsic_memset(
+      /* If we reached here, item_offset is not symbolic
+       * and we don't know what the actual value of it is...
+       *
+       * For now bum_call later expand our simplifier
```
typo

fbrausse commented on src/goto-symex/builtin_functions.cpp:
```diff
@@ -780,246 +1140,181 @@ void goto_symext::intrinsic_memset(
-    // If you're setting a random run of bytes within a struct, best to use
-    // the C implementation.
-    if(is_struct_type(item.object->type) && is_constant_int2t(size))
+    // TODO: Why (X*Y)/Y is not X?
```
overflow maybe? not sure :) can you elaborate?

rafaelsamenezes replied:
Kinda yes. 

However, this operations are happening for offset, which should be safe.

lucasccordeiro commented:
@rafaelsamenezes: What is the status of this PR?

lucasccordeiro commented:
@rafaelsamenezes: could you please address @fbrausse's comments?

rafaelsamenezes committed:
[memset] fixed typo

rafaelsamenezes committed:
[memset] changed simplify calls

rafaelsamenezes committed:
[memset] added std:min usage

rafaelsamenezes committed:
[memset] simplify usage

rafaelsamenezes committed:
[memset] removed uneeded cast

rafaelsamenezes committed:
[memset] updated to new message system

rafaelsamenezes committed:
[memset] code improvments

Added an assertion over float bitwise
operations and replaced a few autos
with concrete types

rafaelsamenezes committed:
[memset] replaced auto with size_t

lucasccordeiro commented:
@rafaelsamenezes: as discussed, we should move forward with this PR :-)

rafaelsamenezes committed:
[memset] improved documentation

rafaelsamenezes committed:
[memset] replaced auto's with types

rafaelsamenezes commented:
This PR:

```
2023-05-26T02:27:40.8875950Z Statistics:          26459 Files
2023-05-26T02:27:40.8878371Z   correct:           14378
2023-05-26T02:27:40.8900102Z     correct true:     8154
2023-05-26T02:27:40.8901427Z     correct false:    6224
2023-05-26T02:27:40.8902410Z   incorrect:            62
2023-05-26T02:27:40.8903417Z     incorrect true:     31
2023-05-26T02:27:40.8904293Z     incorrect false:    31
2023-05-26T02:27:40.8904830Z   unknown:           12019
2023-05-26T02:27:40.8905378Z   Score:             21044 (max: 43043)
```

Master:

```
Statistics:          26459 Files
  correct:           14084
    correct true:     7852
    correct false:    6232
  incorrect:            88
    incorrect true:     42
    incorrect false:    46
  unknown:           12287
  Score:             19856 (max: 43043)
```

From the incorrect solved:

- 6 incorrect false (invalid-deref) became timeouts (goblint-coreutils)
- 15 incorrect in unreach became timeouts
- 4 incorrect true became correct false (in reachability): `aws_array_list_init_dynamic_harness_negated.yml, aws_hash_iter_begin_harness_negated.yml, aws_ring_buffer_init_harness_negated.yml, and ldv-validator-v0.6/linux-stable-4ed3cba-1-100_1a-drivers--usb--serial--qcserial.ko-entry_point.cil.out.yml`
- 2 incorrect became unknown in reachability: still checking why.

Might be worth investigating those 4 incorrect true as this patch shouldn`t improve soundness.

There is two new incorrect results: loops/ludcmp.yml (was correct), and ldv-linux-3.14/linux-3.14_complex_emg_linux-kernel-locking-spinlock_drivers-net-ethernet-dlink-dl2k.cil.yml (was timeout)

lucasccordeiro commented:
Many thanks, @rafaelsamenezes. Can I ask you to fix the code style and rebase this PR with the master?

lucasccordeiro reviewed (approved):
Many thanks, @rafaelsamenezes. Can I ask you to fix the code style and rebase this PR with the master?

rafaelsamenezes committed:
code-style fix

rafaelsamenezes committed:
[memcpy] removed always true ternary

rafaelsamenezes committed:
Update builtin_functions.cpp

rafaelsamenezes committed:
[memset] fixed type so std::min don't complain

rafaelsamenezes committed:
[memset] added a hack for bit_fields

rafaelsamenezes committed:
code-style fix

rafaelsamenezes commented:
Latest PR results:

```
Statistics:          26459 Files
   correct:           14369
     correct true:     8148
     correct false:    6221
   incorrect:            61
     incorrect true:     30
     incorrect false:    31
   unknown:           12029
   Score:             21061 (max: 43043)
```
